pyintegration/deephaven2/table_listener.py

```diff
@@ -0,0 +1,270 @@
+        retain (bool): whether a hard reference to this listener should be maintained to prevent it from being
+            collected, default is True
+        replay_initial_image (bool): False to only process new rows, ignoring any previously existing rows in the Table;
+            True to process updates for all initial rows in the table PLUS all new row changes, default is False
+    """
+
+    _JListenerAdapter = jpy.get_type('io.deephaven.integrations.python.PythonShiftObliviousListenerAdapter')
+    listener = _JListenerAdapter(description, table, retain, implementation)
+    table.j_table.listenForUpdates(listener, replay_initial_image)
+
+
+def python_shift_aware_listener_adapter(table: Table, implementation: Callable, description: str = None,
+                                        retain: bool = True) -> None:
+    """  Constructs the InstrumentedTableUpdateListenerAdapter, implemented in Python, and plugs it into the table's
+    listenForUpdates method.
+
+    Args:
+        table (Table): table to which to listen
+        implementation: the implementation for the InstrumentedTableUpdateListenerAdapter.onUpdate method, and must
+            either be a class with onUpdate method or a callable
+        description (str): a description for the UpdatePerformanceTracker to append to its entry description
+        retain (bool): whether a hard reference to this listener should be maintained to prevent it from being
+            collected, default is True
+    """
+
+    _JListenerAdapter = jpy.get_type('io.deephaven.integrations.python.PythonListenerAdapter')
+    listener = _JListenerAdapter(description, table, retain, implementation)
+    table.j_table.listenForUpdates(listener)
+
+
+class TableListenerHandle:
+    """ A handle for a table listener. """
+
+    def __init__(self, t: Table, listener):
+        """ Creates a new table listener handle.
+
+        Args:
+            t (Table): table being listened to
+            listener: listener object
+        """
+        self.t = t
+        self.listener = listener
+        self.isRegistered = False
+
+    def register(self):
+        """  Register the listener with the table and listen for updates.
+
+        Raises:
+            RuntimeError
+        """
+
+        if self.isRegistered:
+            raise RuntimeError("Attempting to register an already registered listener..")
+
+        self.t.j_table.listenForUpdates(self.listener)
+
+        self.isRegistered = True
+
+    def deregister(self):
+        """ Deregister the listener from the table and stop listening for updates.
+
+        Raises:
+            RuntimeError
+        """
+
+        if not self.isRegistered:
+            raise RuntimeError("Attempting to deregister an unregistered listener..")
+
+        self.t.j_table.removeUpdateListener(self.listener)
+        self.isRegistered = False
+
+
+def _do_locked(f: Callable, lock_type="shared"):
```

Comment:
```suggestion
def _do_locked(f: Callable, lock_type="shared") -> None:
```

Reply:
This function is generally useful.  We should:
1) Consider if it should be public.
2) If it is public, where should it live?

We could drag our feet on a decision if we have a ticket.

Reply:
Yes, it was public in v1 but it is only referenced by listen(). I lack the background knowledge of how such low-level engine integration is used. Also I had trouble figuring out a good place for it and other similar things. Making it private solved the problem and it doesn't preclude us from having a public place for hosting them in the future.